From c2df880c65abbf9d6c92b394b456732501b0cc94 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Thu, 1 Jun 2023 19:23:20 +0200
Subject: [PATCH 0013/1695] core: notif: upgrade to new interrupt framework

Uses main controller ops to call .raise_sgi as the old API functions
itr_xxx() are deprecated.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 6e812e3ed9b06e59d106367bba0107f80991a5c7 core: notif: upgrade to new interrupt framework]
Change-Id: Ib38d614098a18a62d75cc93872ab3de13f1cedbf
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
[etienne: cherry-pick from da637b6b319a81e0107093153b9adb2c821088b3]
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 core/kernel/notif.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/core/kernel/notif.c b/core/kernel/notif.c
index 7ecc246fd..abf9feafa 100644
--- a/core/kernel/notif.c
+++ b/core/kernel/notif.c
@@ -31,6 +31,8 @@ TEE_Result notif_alloc_async_value(uint32_t *val)
 	uint32_t old_itr_status = 0;
 	int bit = 0;
 
+	assert(interrupt_can_raise_pi(interrupt_get_main_chip()));
+
 	old_itr_status = cpu_spin_lock_xsave(&notif_lock);
 
 	if (!alloc_values_inited) {
@@ -93,6 +95,7 @@ out:
 void notif_send_async(uint32_t value)
 {
 	uint32_t old_itr_status = 0;
+	struct itr_chip *itr_chip = interrupt_get_main_chip();
 
 	static_assert(CFG_CORE_ASYNC_NOTIF_GIC_INTID >= GIC_PPI_BASE);
 
@@ -101,7 +104,7 @@ void notif_send_async(uint32_t value)
 
 	DMSG("0x%"PRIx32, value);
 	bit_set(notif_values, value);
-	itr_raise_pi(CFG_CORE_ASYNC_NOTIF_GIC_INTID);
+	interrupt_raise_pi(itr_chip, CFG_CORE_ASYNC_NOTIF_GIC_INTID);
 
 	cpu_spin_unlock_xrestore(&notif_lock, old_itr_status);
 }
@@ -122,6 +125,8 @@ void notif_register_driver(struct notif_driver *ndrv)
 {
 	uint32_t old_itr_status = 0;
 
+	assert(interrupt_can_raise_pi(interrupt_get_main_chip()));
+
 	old_itr_status = cpu_spin_lock_xsave(&notif_lock);
 
 	SLIST_INSERT_HEAD(&notif_driver_head, ndrv, link);
-- 
2.47.2

